### PR TITLE
Implement joint session history

### LINKED ISSUE #11669: Implement joint session history
https://html.spec.whatwg.org/multipage/browsers.html#joint-session-history

The current idea here is to create a `HistoryIterator` that will lazily interleave `Frame` history depending on the `NavigartionDirection`. Each `Frame` entry will now also hold an `Instant` so the order of navigations can be determined.

cc @asajeffrey

### PR DESCRIPTION
**Pull request by cbrewster**
This is cleaned up and should align with the patches on https://github.com/ConnorGBrewster/ServoNavigation/blob/master/notes/notes.pdf
r? @asajeffrey 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11669 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because this is not testable until the History API is added.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11866)

### CODE REVIEW

**Comment by ConnorGBrewster** [comment]
Rebased.

**Comment by jgraham** [comment]
Does this try to implement the spec model? Based on input from Gecko and Blink engineers, it seems that extant implementations are different from the spec in ways that are likely to be difficult to reconcile without changing the spec model to match implementations. In particular the existing implementations store a representation of the frame tree for each position in history, and then diff that tree during navigations to work out which frame to navigate. IMO we should implement that model and change the spec to match.

**Comment by ConnorGBrewster** [comment]
@jgraham it's based off of https://github.com/ConnorGBrewster/ServoNavigation/blob/master/notes/notes.pdf. It tries to stay as close to the current spec as possible. That proposes some changes to the spec to match up with what other browsers do/making traversal homomorphic.

**Comment by asajeffrey** [comment]
@jgraham we tried to make the minimum changes to the spec but be a) as close as possible to existing implementations, and b) vaguely sensible :) We are expecting that spec changes will be a result of this work.

**Comment by jgraham** [comment]
OK, but "as few changes from the spec as possible" should not be a goal. From what I've heard so far it sounds like there is common agreement on how this stuff is implemented, and the spec should reflect that agreement (in this area particularly because other implementations are highly unlikely to change their code to reflect a different model in the spec). From Servo's point of view this doesn't seem like an area where being gratuitously different can buy much in terms of the higher level goals around quality and performance.

**Comment by asajeffrey** [comment]
Agreed, the goals are (a) and (b) and if we can get something that meets this with as little disruption, so much the better.

**Comment by asajeffrey** [comment]
This implementation is based on timestamps and lazily computing the joint session history. I think with the revised spec, we should be able to use an in-memory representation of the jsh, and not have to keep track of timestamps. Not sure which is better.

---

Reviewed 6 of 7 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[components/constellation/constellation.rs, line 220 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLIfsP2HuseFjrhwK1X:-KLIfsP2HuseFjrhwK1Y:1319530124) ([raw file](https://github.com/servo/servo/blob/e5343c0abb6d7f6ad68afad67d05ce01ad45db3e/components/constellation/constellation.rs#L220)):*
> ```Rust
> }
> 
> struct FrameIterator<'a> {
> ```

Rather than keeping two iterators, wouldn't it be easier to have separate `ForwardFrameIterator` and `BackwardFrameInterator` types?

---

*[components/constellation/constellation.rs, line 232 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLIgCrtzaYfYMLWagoR:-KLIgCrtzaYfYMLWagoS:146204144) ([raw file](https://github.com/servo/servo/blob/e5343c0abb6d7f6ad68afad67d05ce01ad45db3e/components/constellation/constellation.rs#L232)):*
> ```Rust
>         match self.direction {
>             TraversalDirection::Back(_) => {
>                 if let Some(entry) = self.current {
> ```

I think this is doing by hand what `once(current).chain(prev)` does.

---

*[components/constellation/constellation.rs, line 266 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLIhQQ3YjgLMhleum8j:-KLIhQQ3YjgLMhleum8k:-961194871) ([raw file](https://github.com/servo/servo/blob/e5343c0abb6d7f6ad68afad67d05ce01ad45db3e/components/constellation/constellation.rs#L266)):*
> ```Rust
>     }
> 
>     fn iter(&self, direction: TraversalDirection) -> FrameIterator {
> ```

I think this would be simpler with two different iterators, since they're different logic.

---

*[components/constellation/constellation.rs, line 364 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLIhcogWncq_4_fW-ep:-KLIhcogWncq_4_fW-eq:-581971522) ([raw file](https://github.com/servo/servo/blob/e5343c0abb6d7f6ad68afad67d05ce01ad45db3e/components/constellation/constellation.rs#L364)):*
> ```Rust
>                 self.stack.iter_mut()
>                           .filter_map(|&mut (frame_id, ref mut iter)| {
>                               iter.peek().cloned().map(|(_, instant)| (frame_id, iter, instant))
> ```

Yay combinators!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866#-:-KLIgr7OCQoczf60hQTc:1584264942)*

**Comment by asajeffrey** [comment]
IRC chat: http://logs.glob.uno/?c=mozilla%23servo&s=27+Jun+2016&e=27+Jun+2016#c463962

TL;DR in-memory vs lazy computation of joint session history?

**Comment by ConnorGBrewster** [comment]
Cleaned up a bit, still a few things I would like to get cleaner, like reducing the number of `Vec` allocations. But this should clean up the `Iterator` stuff a bit.

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/constellation/constellation.rs, line 220 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLIfsP2HuseFjrhwK1X:-KLIzrdqZVc526lhMusk:-256256841) ([raw file](https://github.com/servo/servo/blob/e5343c0abb6d7f6ad68afad67d05ce01ad45db3e/components/constellation/constellation.rs#L220)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Rather than keeping two iterators, wouldn't it be easier to have separate `ForwardFrameIterator` and `BackwardFrameInterator` types?
</details>
I removed the `FrameIterator` altogether, and now store a back and forward stack on the `HistoryIterator`, if it is going forward, the back stack is just an empty vec and vice versa.

---

*[components/constellation/constellation.rs, line 232 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLIgCrtzaYfYMLWagoR:-KLJ1OkfeHj-vVvMat-W:-1147278861) ([raw file](https://github.com/servo/servo/blob/e5343c0abb6d7f6ad68afad67d05ce01ad45db3e/components/constellation/constellation.rs#L232)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> I think this is doing by hand what `once(current).chain(prev)` does.
</details>
Thanks for the tip on `once`, I am not able to remove `FrameIterator`.

---

*[components/constellation/constellation.rs, line 266 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLIhQQ3YjgLMhleum8j:-KLJ1TWTieT-0jK60l8L:719084387) ([raw file](https://github.com/servo/servo/blob/e5343c0abb6d7f6ad68afad67d05ce01ad45db3e/components/constellation/constellation.rs#L266)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> I think this would be simpler with two different iterators, since they're different logic.
</details>
see above.

---

*[components/constellation/constellation.rs, line 364 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLIhcogWncq_4_fW-ep:-KLJ1UWZquLFyJDEa_HR:1185695303) ([raw file](https://github.com/servo/servo/blob/e5343c0abb6d7f6ad68afad67d05ce01ad45db3e/components/constellation/constellation.rs#L364)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Yay combinators!
</details>
:smile:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866#-:-KLJ5HIW8jOHbV1fB9lt:2094688739)*

**Comment by ConnorGBrewster** [comment]
I just cleaned up the `HistoryIterator` quite a bit. Just added a new enum for `HistoryDirection` that holds the corresponding stack type. This gets rid of having an unused empty `Vec`.

**Comment by ConnorGBrewster** [comment]
*[components/constellation/constellation.rs, line 232 \[r8\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLIgCrtzaYfYMLWagoR:-KLJGUpUtS2K9-vf5WLG:977683583) ([raw file](https://github.com/servo/servo/blob/e5343c0abb6d7f6ad68afad67d05ce01ad45db3e/components/constellation/constellation.rs#L232)):*
<details><summary><i>Previously, ConnorGBrewster (Connor Brewster) wrote…</i></summary>
> Thanks for the tip on `once`, I am not able to remove `FrameIterator`.
</details>
s\not\now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866)*

**Comment by ConnorGBrewster** [comment]
I did as much as I could think of to reduce the amount of vec allocations, now there are only 2 vec allocations for `full_frame_tree()`. The pipelines vec can't be switched to a iter because you can't chain while in a for/while loop :cry:. This should be ready for another review pass.

**Comment by asajeffrey** [comment]
OK, I'll have a look tomorrow.

**Comment by asajeffrey** [comment]
Reviewed everything apart from the constellation, i.e. all the easy bits! I'm trying to work out how we can simplify the constellation code.

---

Reviewed 6 of 7 files at r11.
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/msg/constellation_msg.rs, line 246 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLN-9vT8foYicJCh0Mc:-KLN-9vT8foYicJCh0Md:646839341) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/msg/constellation_msg.rs#L246)):*
> ```Rust
> 
> #[derive(Clone, PartialEq, Eq, Copy, Hash, Debug, Deserialize, Serialize)]
> pub enum TraversalDirection {
> ```

Yay for spec-compliant naming!

---

*[components/script_traits/lib.rs, line 551 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLMziWN0Cu1SqgxzmPl:-KLMziWN0Cu1SqgxzmPm:-229850765) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/script_traits/lib.rs#L551)):*
> ```Rust
>     LoadUrl(PipelineId, LoadData),
>     /// Request to traverse the joint session history.
>     TraverseHistory(Option<(PipelineId)>, TraversalDirection),
> ```

Don't need parentheses around `PipelineId`.

---

*[components/script_traits/script_msg.rs, line 77 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLMzbISJBZI4a5FjR3d:-KLMzbISJBZI4a5FjR3e:-229850765) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/script_traits/script_msg.rs#L77)):*
> ```Rust
>     MozBrowserEvent(PipelineId, SubpageId, MozBrowserEvent),
>     /// HTMLIFrameElement Forward or Back traversal.
>     TraverseHistory(Option<(PipelineId)>, TraversalDirection),
> ```

Don't need parentheses around `PipelineId`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866#-:-KLN-JrsXORrXI8HiTDr:698445558)*

**Comment by ConnorGBrewster** [comment]
Finally got back around to this one. I would like to simplify the extra constellation code aswell, it seems like the constellation could grow a lot after this, old pipeline purging, and History API state stuff if we aren't careful.

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/msg/constellation_msg.rs, line 246 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLN-9vT8foYicJCh0Mc:-KLOL791BNNTAu_iSOM4:1203081317) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/msg/constellation_msg.rs#L246)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Yay for spec-compliant naming!
</details>
:tada: 

---

*[components/script_traits/lib.rs, line 551 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLMziWN0Cu1SqgxzmPl:-KLOL8cXNrg1kNsroIux:-871560048) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/script_traits/lib.rs#L551)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Don't need parentheses around `PipelineId`.
</details>
haha woops!

---

*[components/script_traits/script_msg.rs, line 77 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLMzbISJBZI4a5FjR3d:-KLOLAK3phf0TC2dV8qZ:-499146009) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/script_traits/script_msg.rs#L77)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Don't need parentheses around `PipelineId`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866#-:-KLOLCJcmLt_jjVtIGt1:-711445245)*

**Comment by ConnorGBrewster** [comment]
Rebased.

**Comment by ConnorGBrewster** [comment]
Rebased.

**Comment by asajeffrey** [comment]
Reviewed 4 of 7 files at r47, 1 of 1 files at r52, 1 of 2 files at r53.
Review status: 6 of 7 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[components/constellation/constellation.rs, line 311 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8XZvFOktif_o8Ty2a:-KN8XZvFOktif_o8Ty2b:-582551632) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L311)):*
> ```Rust
> type ForwardFrameIterator<'a> = Peekable<Rev<Iter<'a, (PipelineId, Instant)>>>;
> 
> enum HistoryDirection<'a> {
> ```

I don't see why we need this iterator, why not just a forward iterator and a separate backwards iterator?

---

*[components/constellation/constellation.rs, line 654 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8abRz3QMeTFb7F80t:-KN8abRz3QMeTFb7F80u:-1244591942) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L654)):*
> ```Rust
>     // Get a `Vec` of all the frames that are descendants of a root frame. This includes all
>     // active and inactive frames.
>     fn full_frame_tree(&self, frame_id_root: FrameId) -> Vec<FrameId> {
> ```

Would it be better to have two functions, both returning a `Vec<(Instant, PipelineId)>`, one giving the joint session future, and one giving the joint session past and active documents? This vector could then be sorted, which would give the right data for navigating forwards / backwards.

---

*[components/constellation/constellation.rs, line 697 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8dHFF-hnFn9g8Tikv:-KN8dHFF-hnFn9g8Tikw:-610568896) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L697)):*
> ```Rust
>     }
> 
>     // Get an iterator over the joint session history in a given direction.
> ```

If we had the two functions above, I'm not sure we'd need this iterator.

---

*[components/constellation/constellation.rs, line 1572 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8dgZsXmMhYdL1KhHz:-KN8dgZtg9ehGesHlBpv:1756534807) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L1572)):*
> ```Rust
> 
>             let forward = forward_history_iter.next().is_some();
>             let back = back_history_iter.next().is_some();
> ```

These seem like expensive ways to compute forward and back.

---

*[components/constellation/constellation.rs, line 1631 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8e2fsVbJY0oUHNa2-:-KN8e2ftJw59ElGoUcx-:-1820802023) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L1631)):*
> ```Rust
>                             None => 1,
>                         };
>                         traversal_info.insert(frame_id, delta);
> ```

If at this point we updated the `Frame` (e.g. when moving forwards push `current` into `prev` and pop `current` off `next`) and just stored the `frame_id` in a hash table, then we could replace `traverse_frame(FrameId, TraversalDirection)` by `traverse_frame_to(FrameId, PipelineId)`.

---

*[components/constellation/constellation.rs, line 2520 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8g0sQT8sfhVGXFa_P:-KN8g0sRvr9Khhu4vjTm:886497553) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L2520)):*
> ```Rust
>                 if let Some(frame_id) = parent_pipeline.frame {
>                     let mut forward_history_iter = self.history_iterator(frame_id, TraversalDirection::Forward(0));
>                     let mut back_history_iter = self.history_iterator(frame_id, TraversalDirection::Back(0));
> ```

Expensive way to compute `can_go_forward` and `can_go_back`?

---

*[components/script_traits/lib.rs, line 551 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLMziWN0Cu1SqgxzmPl:-KN8gM2mz-kE6idH5wlA:-442767388) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/script_traits/lib.rs#L551)):*
<details><summary><i>Previously, ConnorGBrewster (Connor Brewster) wrote…</i></summary>
> haha woops!
</details>
Shouldn't the `PipelineId` here be a `Frameid`?

---

*[components/script_traits/script_msg.rs, line 77 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLMzbISJBZI4a5FjR3d:-KN8gWzOiWGxkzJbk6d4:-442767388) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/script_traits/script_msg.rs#L77)):*
<details><summary><i>Previously, ConnorGBrewster (Connor Brewster) wrote…</i></summary>
> Done.
</details>
Shouldn't the `PipelineId` here be a `Frameid`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866)*

**Comment by asajeffrey** [comment]
Reviewed 1 of 2 files at r53.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866)*

**Comment by ConnorGBrewster** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/constellation/constellation.rs, line 311 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8XZvFOktif_o8Ty2a:-KNED89XEb3AToaBnCdZ:363052169) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L311)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> I don't see why we need this iterator, why not just a forward iterator and a separate backwards iterator?
</details>
Removed history iterator as we had discussed.

---

*[components/constellation/constellation.rs, line 654 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8abRz3QMeTFb7F80t:-KNEDC6JODFWHJaNOWk3:-499146009) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L654)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Would it be better to have two functions, both returning a `Vec<(Instant, PipelineId)>`, one giving the joint session future, and one giving the joint session past and active documents? This vector could then be sorted, which would give the right data for navigating forwards / backwards.
</details>
Done.

---

*[components/constellation/constellation.rs, line 697 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8dHFF-hnFn9g8Tikv:-KNEDDXsml02M35Uu0TT:-499146009) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L697)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> If we had the two functions above, I'm not sure we'd need this iterator.
</details>
Done.

---

*[components/constellation/constellation.rs, line 1572 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8dgZsXmMhYdL1KhHz:-KNEDECVACBZSxXuYLTc:1935372515) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L1572)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> These seem like expensive ways to compute forward and back.
</details>
I now use the `joint_session_future` and `joint_session_past` and check if they are empty. This may still be too expensive, thoughts?

---

*[components/constellation/constellation.rs, line 1631 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8e2fsVbJY0oUHNa2-:-KNEDPwQt7ZoCMjFJNlY:-499146009) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L1631)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> If at this point we updated the `Frame` (e.g. when moving forwards push `current` into `prev` and pop `current` off `next`) and just stored the `frame_id` in a hash table, then we could replace `traverse_frame(FrameId, TraversalDirection)` by `traverse_frame_to(FrameId, PipelineId)`.
</details>
Done.

---

*[components/constellation/constellation.rs, line 2520 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8g0sQT8sfhVGXFa_P:-KNEDQqS1EY8T_A9PN7g:-1582418045) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L2520)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Expensive way to compute `can_go_forward` and `can_go_back`?
</details>
See above.

---

*[components/script_traits/lib.rs, line 551 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLMziWN0Cu1SqgxzmPl:-KNEDRz6jyK6H5_bmEx8:426842542) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/script_traits/lib.rs#L551)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Shouldn't the `PipelineId` here be a `Frameid`?
</details>
Discussed on IRC. Not changing for now, may be changed in a later PR.

---

*[components/script_traits/script_msg.rs, line 77 \[r17\]](https://reviewable.io:443/reviews/servo/servo/11866#-KLMzbISJBZI4a5FjR3d:-KNEDWcIt26GW5OV6G81:-1582418045) ([raw file](https://github.com/servo/servo/blob/2eed17352da0459fae6eb90fe52a3139af31243c/components/script_traits/script_msg.rs#L77)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Shouldn't the `PipelineId` here be a `Frameid`?
</details>
See above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866)*

**Comment by asajeffrey** [comment]
Reviewed 3 of 7 files at r56, 2 of 2 files at r62, 2 of 2 files at r63.
Review status: 6 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[components/constellation/constellation.rs, line 1572 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8dgZsXmMhYdL1KhHz:-KNEJ1cXculDvXHPhF99:-1316000912) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L1572)):*
<details><summary><i>Previously, ConnorGBrewster (Connor Brewster) wrote…</i></summary>
> I now use the `joint_session_future` and `joint_session_past` and check if they are empty. This may still be too expensive, thoughts?
</details>
OK, we should at least file an issue saying "must get round to fixing this."

---

*[components/constellation/constellation.rs, line 583 \[r63\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEH27hYpyayIsHZ2jG:-KNEH27iLFTMOsBneZd0:-224112408) ([raw file](https://github.com/servo/servo/blob/3ba46f0e1e0819d2df2b6ac1c69cb1a9bc94843b/components/constellation/constellation.rs#L583)):*
> ```Rust
>     }
> 
>     fn joint_session_future(&self, frame_id_root: FrameId) -> Vec<(Instant, FrameId, PipelineId)> {
> ```

Yay, this looks a lot nicer!

---

*[components/constellation/constellation.rs, line 603 \[r63\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEHVUbc-B7TqbQZu71:-KNEHVUcvOvM2R8s5MwD:47772671) ([raw file](https://github.com/servo/servo/blob/3ba46f0e1e0819d2df2b6ac1c69cb1a9bc94843b/components/constellation/constellation.rs#L603)):*
> ```Rust
>             };
>             for &frame_id in &pipeline.children {
>                 future.extend_from_slice(&self.joint_session_future(frame_id));
> ```

Are we creating vectors each time we do a recursive call? Wouldn't it be more efficient to pass in the vector as an argument?

---

*[components/constellation/constellation.rs, line 608 \[r63\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEHkGncxPpeAHyPQ2V:-KNEHkGncxPpeAHyPQ2W:2109687507) ([raw file](https://github.com/servo/servo/blob/3ba46f0e1e0819d2df2b6ac1c69cb1a9bc94843b/components/constellation/constellation.rs#L608)):*
> ```Rust
> 
>         // reverse sorting
>         future.sort_by(|a, b| b.cmp(a));
> ```

Similarly, wouldn't it be more efficient to sort just the once, rather than every recursive call?

---

*[components/constellation/constellation.rs, line 646 \[r63\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEIm_NNsB27jYcrg6u:-KNEIm_NNsB27jYcrg6v:1465146663) ([raw file](https://github.com/servo/servo/blob/3ba46f0e1e0819d2df2b6ac1c69cb1a9bc94843b/components/constellation/constellation.rs#L646)):*
> ```Rust
>     // Get a `Vec` of all the frames that are descendants of a root frame. This includes all
>     // active and inactive frames.
>     fn full_frame_tree(&self, frame_id_root: FrameId) -> Vec<FrameId> {
> ```

Can we get rid of this function? it's only used to remove the joint session future, which we should be able to do without slurping the whole frame tree into a vector.

---

*[components/constellation/constellation.rs, line 2208 \[r63\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEJQZVN4ohZEh5kQqm:-KNEJQZWWhFXqGtv6t3C:214392882) ([raw file](https://github.com/servo/servo/blob/3ba46f0e1e0819d2df2b6ac1c69cb1a9bc94843b/components/constellation/constellation.rs#L2208)):*
> ```Rust
>     }
> 
>     fn remove_forward_history_in_frame_tree(&mut self, frame_id: FrameId) {
> ```

Bikeshed: `clear_joint_session_future`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866)*

**Comment by ConnorGBrewster** [comment]
Review status: 6 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[components/constellation/constellation.rs, line 1572 \[r53\]](https://reviewable.io:443/reviews/servo/servo/11866#-KN8dgZsXmMhYdL1KhHz:-KNEO3dsq-E9j1orfSun:1085483994) ([raw file](https://github.com/servo/servo/blob/6644dda27fe8d807b94d5c80f251d4b040446f06/components/constellation/constellation.rs#L1572)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> OK, we should at least file an issue saying "must get round to fixing this."
</details>
Will do.

---

*[components/constellation/constellation.rs, line 603 \[r63\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEHVUbc-B7TqbQZu71:-KNENzPk0IWwe2poKGEI:-499146009) ([raw file](https://github.com/servo/servo/blob/3ba46f0e1e0819d2df2b6ac1c69cb1a9bc94843b/components/constellation/constellation.rs#L603)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Are we creating vectors each time we do a recursive call? Wouldn't it be more efficient to pass in the vector as an argument?
</details>
Done.

---

*[components/constellation/constellation.rs, line 608 \[r63\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEHkGncxPpeAHyPQ2V:-KNENzlYdXD0yw1NWoyj:-499146009) ([raw file](https://github.com/servo/servo/blob/3ba46f0e1e0819d2df2b6ac1c69cb1a9bc94843b/components/constellation/constellation.rs#L608)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Similarly, wouldn't it be more efficient to sort just the once, rather than every recursive call?
</details>
Done.

---

*[components/constellation/constellation.rs, line 646 \[r63\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEIm_NNsB27jYcrg6u:-KNEO-KqRch_Lt3TEGVB:-2120845828) ([raw file](https://github.com/servo/servo/blob/3ba46f0e1e0819d2df2b6ac1c69cb1a9bc94843b/components/constellation/constellation.rs#L646)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Can we get rid of this function? it's only used to remove the joint session future, which we should be able to do without slurping the whole frame tree into a vector.
</details>
I knew I was forgetting something...

---

*[components/constellation/constellation.rs, line 2208 \[r63\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEJQZVN4ohZEh5kQqm:-KNEO4zRL4Dc-ju1yfee:-499146009) ([raw file](https://github.com/servo/servo/blob/3ba46f0e1e0819d2df2b6ac1c69cb1a9bc94843b/components/constellation/constellation.rs#L2208)):*
<details><summary><i>Previously, asajeffrey (Alan Jeffrey) wrote…</i></summary>
> Bikeshed: `clear_joint_session_future`?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866)*

**Comment by asajeffrey** [comment]
Yay, nearly there, just one more nit.

---

Reviewed 1 of 1 files at r65.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/constellation/constellation.rs, line 592 \[r65\]](https://reviewable.io:443/reviews/servo/servo/11866#-KNEPy0KYJzFczPaj5rZ:-KNEPy0LRr2Pcu9QMXyy:-1111190179) ([raw file](https://github.com/servo/servo/blob/82baa7c125e884b7dd8b3faa598edc16432c4c8b/components/constellation/constellation.rs#L592)):*
> ```Rust
>     }
> 
>     fn get_future_entries(&self, frame_id_root: FrameId, mut future: &mut Vec<(Instant, FrameId, PipelineId)>) {
> ```

Can just make this a Vec rather than a &mut Vec?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11866#-:-KNEQI7y5yPG_RkXqXGF:-1579570305)*

**Comment by asajeffrey** [comment]
My goodness that was a bit epic!

@bors-servo r+

**Comment by asajeffrey** [comment]
```
Tests with unexpected results:
  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/mozbrowser/mozbrowserlocationchange_event.html
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  │ ERROR:js::rust: Error at :0:0: Error: assert_array_equals: property 13, expected false but got true
  └ 

  ▶ Unexpected subtest result in /_mozilla/mozilla/mozbrowser/mozbrowserlocationchange_event.html:
  │ TIMEOUT [expected PASS] Browser API; mozbrowserlocationchange event
  └   → Test timed out
```

**Comment by ConnorGBrewster** [comment]
@bors-servo r=asajeffrey
I was checking the jsh future/past of the root pipeline instead of the mozbrowser iframe pipeline.
